### PR TITLE
[lldb] Use packaging module instead of pkg_resources

### PR DESCRIPTION
**Pull request by JDevlieghere**
Use the packaging [1] module for parsing version numbers, instead of pkg_resources which is distributed with setuptools. I recently switched over to using the latter, knowing it was deprecated (in favor of the packaging module) because it comes with Python out of the box. Newer versions of setuptools have removed `pkg_resources` so we have to use packaging.

[1] https://pypi.org/project/packaging/

### CODE REVIEW

**Review by augusto2112 — APPROVED** [review]

**Review by bulbazord — APPROVED** [review]

**Review by rupprecht — APPROVED** [review]
LGTM, but heads up for @mysterymath if this causes issues w/ Fuchsia's build

**Review by medismailben — APPROVED** [review]

**Comment by mysterymath** [comment]
At least at present, this will probably break Fuchsia's build; the Python we ship includes `pkg_resources` but not `packaging`. Is the default `cpython` build from source intended to build and install `packaging`? Is this something that changed recently?

**Comment by JDevlieghere** [comment]
> At least at present, this will probably break Fuchsia's build; the Python we ship includes `pkg_resources` but not `packaging`. Is the default `cpython` build from source intended to build and install `packaging`? Is this something that changed recently?

I'm not sure but I don't think so. IIUC the expectation that if folks want to use the packaging module, they have to install it themselves (i.e. with pip). I'm not sure all the bots have this package which is the reason I went with the `pkg_resources` approach previously, but that no longer works (i.e. Python 3.12 installed via homebrew no longer has it which is why we've seen folks hitting this).

File: lldb/packages/Python/lldbsuite/test/decorators.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # System modules
 from functools import wraps
-from pkg_resources import packaging
+from packaging.version import parse
```
Comment (labath) [inline lldb/packages/Python/lldbsuite/test/decorators.py:3]:
```suggestion
from packaging import version
```

`parse` sounds too generic to be imported by base name alone

**Review by labath — APPROVED** [review]

File: lldb/test/Shell/helper/build.py [hunk 2]
```diff
@@ -519,9 +519,9 @@ def _find_windows_sdk_in_registry_view(self, view):
 
             # Windows SDK version numbers consist of 4 dotted components, so we
             # have to use LooseVersion, as StrictVersion supports 3 or fewer.
```
Comment (DavidSpickett) [inline lldb/test/Shell/helper/build.py:521]: Does this comment still apply? Are LooseVersion and StrictVersion still a thing we use here?

**Comment by DavidSpickett** [comment]
`packaging` is already installed on Linaro's bots, so we're ok with this.

File: lldb/test/Shell/helper/build.py (hunk 2 above)
Comment (JDevlieghere) [reply lldb/test/Shell/helper/build.py:521]: No, it doesn't. StrictVersion was still being used elsewhere in the file, which I've replaced as well.

**Comment by JDevlieghere** [comment]
GreenDragon now has `packaging` and based on David's message, the Linaro bots do as well. 

@labath Does your ✅ mean the [debian](https://lab.llvm.org/buildbot/#/builders/68) bot has this package installed? 
@mysterymath Can you install this package on the Fuchia bots? I assume you must do something similar for `pexpect`?

**Comment by mysterymath** [comment]
We don't have `pexpect` either; instead the tests that require it are disabled. Our build is as hermetic as possible; even if the package were available in the Python on the bots, we wouldn't be able to use it, since we don't make use of the Python on the bots, for the most part. Instead, we build and ship a CPython from source as part of the bot build.

So far we've avoided providing or requiring any additional Python dependencies, since we have relatively strict requirements for software chain of custody. We'd probably need to figure out a way to build `packaging` from source and to set up a source mirror for it. The previous status quo of checking Python deps into the LLVM monorepo had helped us a lot.

**Comment by mysterymath** [comment]
Did a bit of digging around; it looks like at the very least the Arch Linux python package and the Python docker container don't contain the `packaging` library.

**Comment by labath** [comment]
> @labath Does your ✅ mean the [debian](https://lab.llvm.org/buildbot/#/builders/68) bot has this package installed?

Affirmative.

> Did a bit of digging around; it looks like at the very least the Arch Linux python package and the Python docker container don't contain the `packaging` library. It's included in Arch's `python-pip` package, but having `pip` as a dependency to test LLDB seems odd.

FWIW, there appears to be an actual arch package for this <https://archlinux.org/packages/extra/any/python-packaging/>, so it shouldn't be (I think -- I'm not an Arch user) necessary to install `pip`.

That said, using a package designed (AIUI) for python versions for parsing versions of gcc/clang does strike me as somewhat... unusual, even if it does work, so ***maybe*** there is case to be made for writing something custom (I'm not sure if  we really need anything more elaborate than `tuple([int(part) for part in version.split(".")])`)

**Comment by JDevlieghere** [comment]
> That said, using a package designed (AIUI) for python versions for parsing versions of gcc/clang does strike me as somewhat... unusual, even if it does work, so _**maybe**_ there is case to be made for writing something custom (I'm not sure if we really need anything more elaborate than `tuple([int(part) for part in version.split(".")])`)

Agreed. The first time I took a stab at this, that was the first thing I tried, but I quickly discovered that we had at least one tool (I think it was Python itself?) that contained alphabetical characters (something like `1.2.3rc` or `1.2.3.dev`) and I didn't feel like dealing with all the possible edge cases. We have other packages the test suite relies on (pexpect, psutil, etc) so it seemed reasonable, but if folks feel strongly about it we can maintain our own "version parsing".

**Comment by JDevlieghere** [comment]
> FWIW, there appears to be an actual arch package for this https://archlinux.org/packages/extra/any/python-packaging/, so it shouldn't be (I think -- I'm not an Arch user) necessary to install `pip`.

Homebrew on macOS works similarly, it has a [python-packaging](https://formulae.brew.sh/formula/python-packaging) package that you install with `brew install python-packaging`.

**Comment by mysterymath** [comment]
> We have other packages the test suite relies on (pexpect, psutil, etc) so it seemed reasonable, but if folks feel strongly about it we can maintain our own "version parsing".

Notably the other packages are optional; the lion's share of the test suite can be run without them. Cursorily it looks like this wouldn't be the case here? Admittedly, Fuchsia is the reason at least one of those callouts has been added; we've struggled reigning in LLVM/LLDB's deps due to the manual nature of our build scripts and bureaucratic hassles involved in taking on new deps. But it might be a good idea to have a broader discussion about what the mandatory dep story is/should be for parts of LLVM, rather than doing this piecemeal on patches like this.

**Comment by labath** [comment]
> > That said, using a package designed (AIUI) for python versions for parsing versions of gcc/clang does strike me as somewhat... unusual, even if it does work, so _**maybe**_ there is case to be made for writing something custom (I'm not sure if we really need anything more elaborate than `tuple([int(part) for part in version.split(".")])`)
> 
> Agreed. The first time I took a stab at this, that was the first thing I tried, but I quickly discovered that we had at least one tool (I think it was Python itself?) that contained alphabetical characters (something like `1.2.3rc` or `1.2.3.dev`) and I didn't feel like dealing with all the possible edge cases.

Ah yes, I guess *one* of the versions we are parsing is the version of python itself..

> We have other packages the test suite relies on (pexpect, psutil, etc) so it seemed reasonable, but if folks feel strongly about it we can maintain our own "version parsing".

While I don't think we should be adding deps willy-nilly, I don't think this one is more problematic than others in that they are all available through common package management systems. OTOH, it has two things going against it:
- it's a hard dep breaking the ability to run any test (whereas e.g. pexpect  would only break pexpect-based tests)
- it should be relatively easy to replace

This does not constitute an endorsement of one direction of the other, I'm just thinking out loud.

**Comment by JDevlieghere** [comment]
Thanks for checking @mysterymath. Based on your response it sounds like we have a path forward for the Fuchsia. Please let me know when it's safe to land this. I'll add `packaging` to `requirements.txt` that was added in #94220.

**Comment by mysterymath** [comment]
@JDevlieghere , we just landed a patch that adds packaging for Fuchsia. Should be good to go ahead.

**Comment by aganea** [comment]
This is breaking the github CI as seen here: https://buildkite.com/llvm-project/github-pull-requests/builds/73339#01902659-9fbc-44d2-9a5b-783f5af41b78. Should we add the "packaging" module to https://github.com/llvm/llvm-project/blob/main/lldb/test/requirements.txt ?

**Pull request merged**